gtk: autogenerate GTK resource file from blueprints

jcollie opened the pull request:
Fixes #6760

Eliminate the need for `blueprint-compiler` during build by generating all GTK resource definitions (`.ui` files) from Blueprint files (`.blp` files) during development and checking them into the source repository.

This reduces the number of build dependences needed by people not actively developing the GTK apprt, as well as work around limitations in distros that do not have new enough versions of `blueprint-compiler` packaged (I'm looking at you Debian 12).

A script (accessed by `nix run .#compile-blueprints`) is provided to conveniently recompile all of the blueprints in the repository. A CI job is also added to ensure that the files stay in sync.

mitchellh commented:
These should not be checked into the source tree. I’ll look to see if I can massage this otherwise though. I’m working on eliminating our other build steps too from tarballs. 

Source tree builds will still need the dep.

jcollie committed:
gtk: autogenerate GTK resource file from blueprints

Fixes #6760

Remove all GTK resource definitions (`.ui` files) from the repo.
Building from a git checkout requires `blueprint-compiler` version
0.16.0 at build time to compile the Blueprints (`.blp` files) into GTK
resource definitions.

A script (accessed by `nix run .#compile-blueprints`) is provided
to compile all of the blueprints in the repository for inclusion in
generated tarballs. A CI job is also added to ensure that the Blueprint
files are valid and formatter properly.

Workflows for tag and tip will generate source tarballs that contain
generated GTK resource definitions for use on distributions that do not
provide a new enough `blueprint-compiler` (I'm looking at you Debian
12).

Building from generated tarballs can use the pre-generated GTK resource
definitions by adding the `-fno-sys=blueprint-compiler` flag to the
Zig build.

jcollie committed:
gtk/macos: don't add anything to GI_TYPELIB_PATH on non-Linux

jcollie committed:
gtk: add GI_TYPELIB_PATH env var when compiling blueprints

jcollie committed:
snap: use Nix to compile the blueprints before building

jcollie committed:
snap: don't try and use blueprint-compiler

mitchellh reviewed (changes requested):
Going to mark this as "requesting changes" just so it doesn't get merged.

As noted in the previous comment, there's some changes I want to make here. The improvements made in subsequent commits are good but still not completely in the direction I'm imagining. 

I expect I'll use some of this work but step 1 I want to shore up our source tarball generation into the Zig build system. I also want to rely less on `nix run` and more on `zig build` (which can of course use Nix dependencies). 

It also _looks like_ there are some unrelated changes mixed into this PR? I'm not sure, its hard to tell. If so, please pull those out into separate PRs.

I don't expect to merge this.

mitchellh commented:
Closing in favor of #6822. There seems to be some useful things in here, so if you want to bring back those parts on top of the latest changes, feel free!
